### PR TITLE
Move file loading and processing async in SANS

### LINKED ISSUE #27216: GUI Not generating SANS States correctly
SANS States are not being generated correctly. This manifests itself as a warning stating that a user's file is not found, despite being found.

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
Move file finding to be synchronous as we need the result before
processing or loading data. Instead we now run the load or process
method async within the worker to avoid blocking the main thread

**To test:**
- Using the LOQ data set:
- Attempt to load or process a data set
- An error previously would appear and pressing ok and process again would work
- Now it should process first time
- Try various combo's with load and process selected/all

Fixes #27216 .

*This does not require release notes* because this issue was introduced and resolved before a release

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/SANS/sans/gui_logic/models/batch_process_runner.py [hunk 1]
```diff
@@ -65,7 +77,18 @@ def _process_states_on_thread(self, states, use_optimizations, output_mode, plot
...
+    def _load_workspaces_on_thread(self, selected_rows, get_states_func, get_thickness_for_rows_func):
+        get_thickness_for_rows_func()
+        # The above must finish before we can call get states
+        states, errors = get_states_func(row_index=selected_rows)
+
+        for row, error in errors.items():
+            self.row_failed_signal.emit(row, error)
+
+        if not states:
+            self.row_processed_signal.emit(None)
```
Comment (Mantid-Matthew) [inline scripts/SANS/sans/gui_logic/models/batch_process_runner.py]: Elsewhere this signal is emitted with 3 arguments. Will it work being emitted with one here?

**Review by Matthew-Andrew — CHANGES_REQUESTED** [review]
Code changes look good and the double error is gone. A few side issues however:

-  If a run is not found this error is not propagated up correctly, i.e. does not turn the row blue e.c.t.
-  The progress bar at the bottom of the gui does not reset correctly
-  Should post loading be indicated by green which could imply reduced?
-   If you delete some rows during processing you get an unhandled exception.

**Comment by DavidFair** [comment]
- If a run is not found this error is not propagated up correctly, i.e. does not turn the row blue e.c.t.
This should fixed
- The progress bar at the bottom of the gui does not reset correctly
Fixed

- Should post loading be indicated by green which could imply reduced?
I'm tempted to leave this to user feedback, as if it causes confusion (green = processed rather than finished doing) we can change it
- If you delete some rows during processing you get an unhandled exception.
I'm going to defer this to the next release, as there are multiple ways to delete a row during processing, so just stopping the buttons working will not prevent this happening.

**Review by Matthew-Andrew — CHANGES_REQUESTED** [review]
Looks good overall the failing to find a file is now notified properly :+1: 

If I add some rows click process all and the add some more rows and click process all the progress bar behaves strangely however often reaching zero early or increasing in odd increments.

**Comment by DavidFair** [comment]
Good spot on the progress bar, I've added a fix and a unit test for this

**Review by Matthew-Andrew — APPROVED** [review]
Works for me now :+1:

**Pull request merged**